### ISSUE TITLE
Feature request: packages as git submodules

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1950]
Here is a suggestion: instead of organizing the `packages` directory as:
```
Macaulay2/packages/
...
├── Matroids.m2
├── Matroids
│   └── SmallMatroids.txt
...
```
let's allow having packages contained in another git repository and allow such an organization:
```
Macaulay2/packages/
...
├── Matroids
│   ├── Matroids.m2
│   └── SmallMatroids.txt
...
```
This wouldn't need to happen for every package at once, but all new packages, and optional for updated packages.

This has many advantages:
- developers (e.g. @giovannistagliano) can continuously update their packages
- users can get access to updated versions of a package immediately (e.g. `pullPackage`)
- users can get access to new packages before a release (e.g. `clonePackage`)
- github actions can run the checks of the individual package on each commit quickly
- github actions can generate examples and documentation and upload it on Github Pages
- before each release, the developers open a pull request to update the submodule hash, at which point all packages are checked together in order to guarantee interoperability
- this method guarantees that old packages stay operational, because we still run all checks on all packages.
- it would also reduce messiness in the packages directory.

Possible extensions:
- we can potentially implement `pushPackage`, which would automate updating a package.
- allow packages names like `Numerical/SchubertCalculus.m2`, so multiple packages can be grouped together under the `Numerical` umbrella and a single repository can contain all of them. Updating those packages at the same time would be simpler this way.
- if Macaulay2 can be made to understand git, then `PackageImports` and `PackageExports` can take an argument like `"LocalRings" => "mahrud/LocalRings.m2 @ [commit hash]"`, which means depend at least on that commit or better.

This is basically universal in all other scripting programming languages with support for packages or modules. We can learn a lot from systems like `brew`, `pypi`, `npm`, [GAP-packages](https://gap-packages.github.io/), JuliaHub, etc.

**Comment by mahrud** [comment]
Interesting perspective here: https://github.com/joelparkerhenderson/monorepo_vs_polyrepo

Instead of each package being a separate repository, I wonder if it's better to associate to each package a fork of M2 (e.g. `mahrud/LocalRings`) that serves as the "upstream" for that package only. That branch can then have tags like `LocalRings-v1.0` which points to the state of M2 when that package was released. This would make is super easy to benchmark how changes since that particular version have affected the performance of a package.

**Comment by mahrud** [comment]
Just bumping this up in case anyone would like to read this more carefully or comment before there's an implementation.

cc: @mikestillman, @DanGrayson, @antonleykin

**Comment by DanGrayson** [comment]
Re: "developers (e.g. @giovannistagliano) can continuously update their packages"

They can already do that, on their own fork.

Re: "users can get access to updated versions of a package immediately"

A pull request for a simple update to a package takes only days to review.  (Are you proposing that we don't test the updates to packages before users get them?  That would not be good.)

**Comment by mahrud** [comment]
> A pull request for a simple update to a package takes only days to review.

It takes a few days to review + 6 months to make it to the next version of M2 + some time before users actually update their M2. Have you seen the [ad-hoc ways](https://github.com/Macaulay2/M2/blob/729dcedb9437682f631c99e1473e47732dc06d1f/M2/Macaulay2/packages/MultiprojectiveVarieties.m2#L25-L35) people are using to update individual packages quickly? Waiting for you to make a new release isn't fast enough.

**Comment by DanGrayson** [comment]
That code Giovanni wrote is certainly ad hoc, and something should be done, but doing it can be done easily without making each package into a submodule.  One could write a new function called `getPackage`, so that `getPackage "Cremona"` would download the package from the master branch of the Macaulay2 git repository into the `code` subdirectory of the user application directory (which is on the path, and is the place intended for downloaded packages to be deposited), and then it would run `installPackage "Cremona"`.

We could also invent a notation for the package options to `newPackage` for encoding a minimum required version, signaling an error if the minimum is not satisfied.

**Comment by mahrud** [comment]
I think you missed the entire point of this proposal, which is to move Macaulay2 package development away from being a [closed platform](https://en.wikipedia.org/wiki/Closed_platform) in favor of an [open platform](https://en.wikipedia.org/wiki/Open_platform):

> This is basically universal in all other scripting programming languages with support for packages or modules. We can learn a lot from systems like `brew`, `pypi`, `npm`, [GAP-packages](https://gap-packages.github.io/), JuliaHub, etc.

This could be done while still ensuring that new versions of packages remain compatible with previous packages.

**Comment by DanGrayson** [comment]
Okay, but I'm bemused that, if that's the entire point of the proposal, this is the first time it is mentioned.

In what sense is Macaulay2 not already an open platform?  Packages can already be distributed by 3rd parties -- the user uses `installPackage` to install them locally after downloading them and obtains the full benefit.

**Comment by mahrud** [comment]
> In what sense is Macaulay2 not already an open platform? Packages can already be distributed by 3rd parties -- the user uses `installPackage` to install them locally after downloading them and obtains the full benefit.

In the sense that you are the judge of which version of a package the users are "granted" access to, unless they can email the authors for an updated version package and wait for a response.

**Comment by DanGrayson** [comment]
No, a package author can advertise and distribute a package independently, without submitting it to us at all.

**Comment by mahrud** [comment]
I'll just quote myself again: having a way to get packages from git repositories automatically is ...
> ... basically universal in all other scripting programming languages with support for packages or modules. We can learn a lot from systems like brew, pypi, npm, GAP-packages, JuliaHub, etc.

**Comment by mahrud** [comment]
There are also several other benefits of this proposal which @antonleykin and others discussed last October, including easier management of issues and pull requests related to a specific package.

For instance a repository for NumericalAlgebraicGeometry and related packages would host all issues related to those packages, rather than being lost in some 600+ issues on this repository, and people can make pull requests there that can be reviewed by the authors, and when a large enough chunk of changes are ready, the authors can make a pull request to update the submodule hashes here.

**Comment by DanGrayson** [comment]
Okay -- for "a way to get packages from git repositories automatically", it would be easy to add an optional named argument to `getPackage` (proposed above) that specifies the source location of the package, which could be a git repository or a tar file, etc. 

And for easier management of package development, having a submodule, as you propose, is a good idea.  (A simple change to `loadPackage` would be to have it look both for `Foo.m2` and for `Foo/Foo.m2`.)

(The two directions seem independent.)

**Comment by DanGrayson** [comment]
Mahrud, let's consider an alternative to changing the way a package is organized.  Instead of each package being a submodule, let's think of each group of related packages as being a submodule, not necessarily of the directory `packages`,  but of a new directory called something like `packageGroups`, which could be a superdirectory of `packages`.  This will be better -- look at Frank Schreyer's packages -- 10 of them have to do with constructing things randomly.  Also there are several packages related to numerical algebraic geometry, organized by Anton Leykin.

**Comment by DanGrayson** [comment]
... An advantage of doing it this way is that we don't need to modify `installPackage` and `loadPackage` -- just add all the subdirectories of `packageGroup` to the path.

**Comment by antonleykin** [comment]
It looks like we all agree that separating out (git-wise) packages as submodules is a good idea. I also agree that a one-to-one correspondence of packages and submodules is not absolutely necessary.

I'd be happy to provide `NumericalAlgebraicGeometry-group` as a guinea pig.

**Comment by mahrud** [comment]
I don't think there's a need for new terminology like `packageGroup` or changing the path. With some minor changes in `loadPackage` it should be possible to have something like the following:
```
Macaulay2/packages/
...
├── NumericalAlgebraicGeometry
│   ├── NumericalAlgebraicGeometry.m2
│   ├── NumericalLinearAlgebra
│   │   └── NumericalLinearAlgebra.m2
│   ├── NumericalCertification
│   │   └── NumericalCertification.m2
│   ...
...
```
Now `NumericalAlgebraicGeometry` can be a submodule.

**Comment by DanGrayson** [comment]
You're forgetting about the changes required to `installPackage`, too.  I think the other way is much simpler : package groups.

**Comment by mahrud** [comment]
I'm not forgetting anything. I know because I've had a draft of this for a while and it works fine. `installPackage` simply calls `loadPackage` at the beginning, so the changes are limited to that function.

**Comment by DanGrayson** [comment]
A disadvantage of packages as submodules, which cropped up in our discussion at the internals meeting yesterday, is this:  Suppose we have to edit a package -- this can happen when the package is published in JSAG and we want to certify it, or when we make a breaking change to the core of Macaulay2 and we want to fix the packages that are broken.  To edit a package that is part of a submodule, we have to make a fork of the submodule, edit it, and redirect Macaulay2's pointer to the submodule so it refers to the fork.

If a package author wants to share development of the package, they can do so with a fork of the M2 repository quite easily.  Co-authors can have discussions, issues, and pull requests on that fork.  So what actually is the advantage of packages as submodules?

**Comment by DanGrayson** [comment]
- 

> I'm not forgetting anything. I know because I've had a draft of this for a while and it works fine. `installPackage` simply calls `loadPackage` at the beginning, so the changes are limited to that function.

Well, you might be right, I haven't looked at the code recently, but `installPackage` has to also install the package by copying all the files in it -- doesn't it get confused when `Foo.m2` isn't where it is expected to be?

**Comment by DanGrayson** [comment]
I advocate a conservative approach to the structure of a package.  Currently it is a file `Foo.m2` with an optional sibling directory `Foo`.  We can preserve that easily.

**Comment by antonleykin** [comment]
I'd support a conservative (or rather minimalistic yet _non-empty_) approach.
How about we provide a way to maintain a package (or a group of packages) as
- a git submodule
- a git fork (of Macaulay2/M2)

and also a way to get recent updates (e.g., via `getPackage` or `installPackage`) in both cases.

Changing the file structure would not be necessary.

**Comment by DanGrayson** [comment]
The conservative approach, preserving the current package layout, would be to introduce package groups, laid out like this:
```
Macaulay2/packageGroups/
...
├── NumericalAlgebraicGeometry
│   ├── NumericalAlgebraicGeometry.m2
│   ├── NumericalLinearAlgebra.m2
│   ├── NumericalLinearAlgebra/
│   ├── NumericalCertification.m2
│   ├── NumericalCertification/
│   ...
...
```

And it offers advantages to those who are developing their packages in their own repositories -- when including the package in Macaulay2, the history is included.

**Comment by mahrud** [comment]
Why does `packageGroups` it need to be a separate directory? You can either add an option to `newPackage` or maybe create a list `=distributed-package-groups`.

**Comment by DanGrayson** [comment]
A list of distributed package groups is a good idea, thanks!  Then everything can reside in the packages directory.

**Comment by mahrud** [comment]
I agree that updates to multiple packages at the same time are harder to deal with as submodules, but there are solutions to make that easier.

That said, I think for now only changing the directory structure and introducing subpackages (rather than thinking of them as package groups) is probably enough changes.

Also, I think the package interdependencies are a larger problem in M2 that need a solution of their own, particularly for dependency loops and packages with "Old" prefix.

**Comment by DanGrayson** [comment]
I'm pretty sure a package dependency loop is signalled with an error.

One good thing about introducing package groups is the that group might contain just one package, so we satisfy two desires with one change.

**Comment by mahrud** [comment]
> I'm pretty sure a package dependency loop is signalled with an error.

It shouldn't be! Packages should be allowed to have co-dependencies so authors don't have to put functions in places they don't belong to!

**Comment by DanGrayson** [comment]
> > I'm pretty sure a package dependency loop is signalled with an error.
> 
> It shouldn't be! Packages should be allowed to have co-dependencies so authors don't have to put functions in places they don't belong to!

I don't understand.  Declaring that package X depends on Y implies that Y should be loaded first, and then X should be loaded.  Hence a loop is impossible.

**Comment by DanGrayson** [comment]
I just did a test and got a surprising result -- these two packages give no error, even back to M2 1.12:
```m2
newPackage ( "Foo",
     PackageImports => { "Bar" }
     )

-----------------------------------------------------------------------------

newPackage ( "Bar",
     PackageImports => { "Foo" }
     )
```

Will fix.
